client/src/main/java/org/spine3/validate/ConstraintViolations.java

```diff
@@ -128,4 +137,89 @@ public static String toText(String format, Iterable<ConstraintViolation> violati
         }
         return resultBuilder.toString();
     }
+
+    /**
+     * A helper class for building exceptions used to report invalid {@code Message}s,
+     * which have fields that violate validation constraint(s).
+     *
+     * @param <E> type of {@code Exception} to build
+     * @param <M> type of the {@code Message}
+     * @param <C> type of the {@linkplain MessageClass MessageClass} of {@code |M|}.
+     * @param <R> type of an error code to use for error reporting; must be a Protobuf enum value
+     */
+    @Internal
+    public abstract static class ExceptionBuilder<E extends Exception,
+                                                  M extends Message,
+                                                  C extends MessageClass,
+                                                  R extends ProtocolMessageEnum> {
+
+        private final Iterable<ConstraintViolation> constraintViolations;
+        private final M message;
+
+        /**
+         * Creates an {@code ExceptionBuilder} instance for a given message and
+         * constraint violations,
+         *
+         * @param message              an invalid event message
+         * @param constraintViolations constraint violations for the event message
+         */
+        protected ExceptionBuilder(M message,
+                                   Iterable<ConstraintViolation> constraintViolations) {
+            this.constraintViolations = constraintViolations;
+            this.message = message;
+        }
+
+        /**
+         * Obtains a {@code MessageClass} for an invalid {@code Message}.
+         */
+        protected abstract C getMessageClass();
+
+        /**
+         * Obtains an error code to use for error reporting.
+         */
+        protected abstract R getErrorCode();
+
+        protected abstract String getErrorText();
+
+
+        protected abstract Map<String, Value> getMessageTypeAttribute(Message message);
+
+        protected abstract E createException(String exceptionMsg, M message, Error error);
+
+        private String formatExceptionMessage() {
+            return String.format("%s. Message class: %s. " +
```

Comment:
There are 16 more usages of `String.format()` (20 in total)  in our code. 
And 22 statically imported use-cases. 

Let's discuss and fix as a separate PR.